pkg/utils/db.go

```diff
@@ -473,3 +473,45 @@ func ExtractTiDBVersion(version string) (*semver.Version, error) {
 	rawVersion = strings.TrimPrefix(rawVersion, "v")
 	return semver.NewVersion(rawVersion)
 }
+
+// AddGSetWithPurged is used to handle this case: https://github.com/pingcap/dm/issues/1418
+// we might get a gtid set from Previous_gtids event in binlog, but that gtid set can't be used to start a gtid sync
+// because it doesn't cover all gtid_purged. The error of using it will be
+// ERROR 1236 (HY000): The slave is connecting using CHANGE MASTER TO MASTER_AUTO_POSITION = 1, but the master has purged binary logs containing GTIDs that the slave requires.
+// so we add gtid_purged to it.
+func AddGSetWithPurged(ctx context.Context, gset gtid.Set, conn *sql.Conn) (gtid.Set, error) {
```

Comment:
Why do we use `sql.Conn` instead of `sql.DB` here?

Reply:
some callers only hold sql.Conn not sql.DB

Reply:
in line 108 of pkg/v1dbschema/schema.go (below file). I'll close connection in next commit